FEATURE: Add `onAfterHandle` to CommandHookInterface

mficzel opened the pull request:
Followup to #5336

This allows to return additional commands after the just executed one was handled. The onAfterHandle method gets the processed command plus all resulting events.

Example use cases: Add automatic translations to a freshly created node variant. Or create translations in a review workspace when changes are published to the live workspace.

Resolves: #5469 
<!-- 
    Thanks for your contribution, we appreciate it!

    The first section should explain briefly what is changed. 
    Some examples are always nice to showcase the use. 
    The content will be used in the change-logs and addresses 
    developers working with Neos.

    If there are issues regarding the topic of your PR link 
    them here as `related:` or `resolved:`
-->

**Upgrade instructions**

<!-- 
    Add upgrade instructions for breaking changes. 
    Explain who is affected, what has to be adjusted  
-->

**Review instructions**

<!-- 
    If your change is not explained fully by the first section you can 
    add more details here to help the reviewers understand the change.
    We have to understand what you did, why you did it and how we can 
    verify it works correctly and does no harm.
-->

**Checklist**

- [x] Code follows the PSR-2 coding style
- [x] Tests have been created, run and adjusted as needed
- [x] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

mficzel committed:
TASK: Add onAfterHandle to CommandHookInterface

This allows to return additional Commands after the just executed ones were projected.
Example. Add automatic translations to a freshly created node variant.

mhsdesign commented on Neos.ContentRepository.Core/Classes/ContentRepository.php:
```diff
@@ -130,6 +131,7 @@ public function handle(CommandInterface $command): void
                     throw $concurrencyException;
                 }
             }
+            $this->commandHook->onAfterHandle($command);
```
this line must be executed _after_ belows `if ($fullCatchUpResult->hadErrors()) {` as other-wise the cr is not up to date. Though im not sure if the hook should be run in the `finally` block if there was an exception like `$concurrencyException` ... probably not. So we need to change the `finally` block to be `catch (\Throwable)` and and run the `onAfterHandle` only if that was not the case.

mficzel replied:
Yap ... I will take a deeper look. Just wanted to get it working first.

mficzel replied:
Moved the hook to the end of the finally block. In case the `$fullCatchUpResult->hadErrors()` this is never reached because an error is thrown before. That way the behavior is equal regardless wether Events or a Generator are beeing committed.

mficzel commented:
Note: After this is merged the existing use cases should be updated as well and return empty commands onHandle. 

1. https://github.com/neos/form-builder/blob/main/Classes%2FCommandHook%2FUniqueIdentifierCommandHook.php

mficzel committed:
TASK: Add the processed events as argument to the CommandHookInterface->onAfterHandle()

mhsdesign committed:
TASK: Introduce `t_extensibility` CR and `CommandHookTest`

mhsdesign committed:
TASK: Do not run `onAfterHandle` during exceptions after control flow aware commands

mhsdesign reviewed (commented):
Events is not API! And contains the Decorated Event which is not api either and that makes sense as this is write only ... also the time stamps etc might be cool to have

The catchup hook or projections get:

`onBeforeEvent(EventInterface $eventInstance, EventEnvelope $eventEnvelope)`

But we dont have the `EventEnvelope` at hand at the moment as this is only a thing when fetching the events again...

as we also talked about performance maybe we dont want to have each hook load the events now ... but we could hand over them the event stream to look into it ... thought the stream is wayyy to low level and requires the internal EventNormaliser too ...

Renaming the internal `Events` to `EventsToWrite` and then introduce a `Events` thing with only `EventInterface` might be a way? For accessing its eventMetaData one would still have to use a catchup hook ...


Otherwise i really like the idea and added some cool tests 🧪 :D

mhsdesign committed:
TASK: Introduce new simple API `PublishedEvents` class for use in command hook

We dont want to use the `Events` class as they are only needed for the write site and internal.

And instead of renaming the internal `Events` class https://github.com/neos/neos-development-collection/pull/5486 just to create a new plain `Events` class.

mhsdesign reviewed (approved):
The tests show that it works thanks a lot for this extensibility point! :)
We discussed this feature several times in a bigger group and all think its of good value!

dlubitz commented:
> Note: After this is merged the existing use cases should be updated as well and return empty commands onHandle.
> 
> 1. https://github.com/neos/form-builder/blob/main/Classes%2FCommandHook%2FUniqueIdentifierCommandHook.php

There are more usages already. Not sure I can remember all of them :/ e.g:
* https://github.com/webexcess/openstreetmap/pull/32

Here is a github wide search: https://github.com/search?q=Neos%5CContentRepository%5CCore%5CCommandHandler%5CCommandHookInterface&type=code

dlubitz reviewed (approved):
Looks good by reading.

mficzel commented:
Just for documentation ... here is the pr to the form builder that is required after this https://github.com/neos/form-builder/pull/146

dlubitz commented:
I prepared this one:
https://github.com/webexcess/openstreetmap/pull/33
